notebooks/search/05-query-rules.ipynb

```diff
@@ -0,0 +1,593 @@
+        "for doc in docs:\n",
+        "    actions.append({\"index\": {\"_index\": \"products_index\", \"_id\": doc[\"id\"]}})\n",
+        "    actions.append(doc[\"content\"])\n",
+        "client.bulk(index=\"products_index\", operations=actions)\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "985c9992",
+      "metadata": {},
+      "source": [
+        "## Search the test data\n",
+        "\n",
+        "First, let's search our data for a reliable wireless charger.\n",
+        "\n",
+        "Before we search our data, we'll define some convenience functions to output the raw JSON responses from Elasticsearch into more human-readable format."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "54b40035",
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "def pretty_response(response):\n",
+        "    for hit in response['hits']['hits']:\n",
+        "        id = hit['_id']\n",
+        "        score = hit['_score']\n",
+        "        name = hit['_source']['name']\n",
+        "        description = hit['_source']['description']\n",
+        "        price = hit[\"_source\"][\"price\"]\n",
+        "        currency = hit[\"_source\"][\"currency\"]\n",
+        "        plug_type = hit[\"_source\"][\"plug_type\"]\n",
+        "        voltage = hit[\"_source\"][\"voltage\"]\n",
+        "        pretty_output = (f\"\\nID: {id}\\nName: {name}\\nDescription: {description}\\nPrice: {price}\\nCurrency: {currency}\\nPlug type: {plug_type}\\nVoltage: {voltage}\\nScore: {score}\")\n",
+        "        print(pretty_output)\n",
+        "\n",
+        "def pretty_ruleset(response):\n",
+        "    print(\"Ruleset ID: \" + response['ruleset_id'])\n",
+        "    for rule in response['rules']:\n",
+        "        rule_id = rule['rule_id']\n",
+        "        type = rule['type']\n",
+        "        print(f\"\\nRule ID: {rule_id}\\n\\tType: {type}\\n\\tCriteria:\")\n",
+        "        criteria = rule['criteria']\n",
+        "        for rule_criteria in criteria:\n",
+        "            criteria_type = rule_criteria['type']\n",
+        "            metadata = rule_criteria['metadata']\n",
+        "            values = rule_criteria['values']\n",
+        "            print(f\"\\t\\t{metadata} {criteria_type} {values}\")\n",
+        "        ids = rule['actions']['ids']\n",
+        "        print(f\"\\tPinned ids: {ids}\")"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "ed5c572a",
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "response = client.search(index=\"products_index\", body={\n",
+        "    \"query\": {\n",
+        "      \"multi_match\": {\n",
+        "          \"query\": \"reliable wireless charger for iPhone\",\n",
+        "          \"fields\": [ \"name^5\", \"description\" ]\n",
+        "      }\n",
+        "    }\n",
+        "})\n",
```

Comment:
I have a small nit here. The Elasticsearch Python client, starting with version 8.0, supports parameters for the first level of body keys. Which means you can write:

```python3
response = client.search(index="products_index", query={"multi_match": ...})
```

The `body` form you're using here is fine (it was deprecated in 8.0 but that deprecation will go away in 8.11 or maybe earlier) but I think it would be nice to be consistent throughout the notebook. So I would either switch to `body` for `client.query_ruleset.put` or switch to `query` for `client.search`. I think it would make the `query_ruleset` call feel less magic.

Reply:
Done, thanks!